gstools/covmodel/fit.py

```diff
@@ -0,0 +1,379 @@
+    pcov : :class:`numpy.ndarray`
+        The estimated covariance of `popt` from
+        :any:`scipy.optimize.curve_fit`.
+        To compute one standard deviation errors
+        on the parameters use ``perr = np.sqrt(np.diag(pcov))``.
+    r2_score : :class:`float`, optional
+        r2 score of the curve fitting results. Only if return_r2 is True.
+
+    Notes
+    -----
+    You can set the bounds for each parameter by accessing
+    :any:`CovModel.set_arg_bounds`.
+
+    The fitted parameters will be instantly set in the model.
+    """
+    # preprocess selected parameters
+    var_last = False
+    for par in para_select:
+        if par not in model.arg_bounds:
+            raise ValueError(
+                "fit: unknow parameter in selection: {}".format(par)
+            )
+        if not isinstance(para_select[par], bool):
+            if par == "var":
+                var_last = True
+                var_tmp = float(para_select[par])
+            else:
+                setattr(model, par, float(para_select[par]))
+            para_select[par] = False
+    # set variance last due to possible recalculations
+    if var_last:
+        model.var = var_tmp
+    # remove those that were set to True
+    para_select = {k: v for k, v in para_select.items() if not v}
+
+    # handling the sill
+    sill = None if (isinstance(sill, bool) and sill) else sill
+    if sill is not None:
+        sill = model.sill if isinstance(sill, bool) else float(sill)
+        constrain_sill = True
+        sill_low = model.arg_bounds["var"][0] + model.arg_bounds["nugget"][0]
+        sill_up = model.arg_bounds["var"][1] + model.arg_bounds["nugget"][1]
+        if not sill_low <= sill <= sill_up:
+            raise ValueError("fit: sill out of bounds.")
+        if "var" in para_select and "nugget" in para_select:
+            if model.var > sill:
+                model.nugget = model.arg_bounds["nugget"][0]
+                model.var = sill - model.nugget
+            else:
+                model.nugget = sill - model.var
+        elif "var" in para_select:
+            if model.var > sill:
+                raise ValueError(
+                    "fit: if sill is fixed and variance deselected, "
+                    + "the set variance should be less than the given sill."
+                )
+            para_select["nugget"] = False
+            model.nugget = sill - model.var
+        elif "nugget" in para_select:
+            if model.nugget > sill:
+                raise ValueError(
+                    "fit: if sill is fixed and nugget deselected, "
+                    + "the set nugget should be less than the given sill."
+                )
+            para_select["var"] = False
+            model.var = sill - model.nugget
+        else:
+            # deselect the nugget, to recalculate it accordingly
+            # nugget = sill - var
+            para_select["nugget"] = False
+    else:
+        constrain_sill = False
```

Comment:
Would it make sense to outsource stuff like this to helper functions for better readability?

Reply:
Done. Thanks for the reminder!